devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+import sympy
+import numpy as np
+
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
+
+__all__ = ['Differentiable']
+
+
+class Differentiable(FrozenExpr):
+    """
+    This class represents Devito differentiable objects such as
+    sum of functions, product of function or FD approximation and
+    provides FD shortcuts for such expressions
+    """
+    _op_priority = 100.0
+    is_Function = True
+    is_Indexed = False
+
+    def __new__(cls, *args, **kwargs):
+        return sympy.Expr.__new__(cls, *args)
+
+    def __init__(self, expr, **kwargs):
+        from devito.finite_differences.finite_difference import generate_fd_functions
+        self.expr = expr.expr if isinstance(expr, Differentiable) else expr
```

Comment:
I think this is the source of most of the issues that I commented below (in fact, this one that you're reading is the very last comment that I wrote in this buck). Why do we need ``expr`` to be stored as a field? what's the symbolic difference between ``self``  and ``expr`` at this point?